Possible race condition in geometry package

When the tests are run with `-race`, a race condition is detected in the `geometry` package:

```
==================
WARNING: DATA RACE
Read at 0x00c000138120 by goroutine 106:
  reflect.typedmemmove()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/mbarrier.go:193 +0x0
  reflect.packEface()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/reflect/value.go:135 +0xa4
  reflect.valueInterface()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/reflect/value.go:1520 +0x148
  reflect.Value.Interface()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/reflect/value.go:1490 +0x54
  github.com/alecthomas/repr.asTime()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:347 +0x58
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:241 +0xf30
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:311 +0xad0
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:271 +0x18d8
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:311 +0xad0
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:271 +0x18d8
  github.com/alecthomas/repr.(*Printer).reprValue()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:311 +0xad0
  github.com/alecthomas/repr.(*Printer).Print()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:141 +0xc0
  github.com/alecthomas/repr.String()
      /Users/twp/pkg/mod/github.com/alecthomas/repr@v0.3.0/repr.go:356 +0x1d8
  github.com/alecthomas/assert/v2.objectsAreEqual()
      /Users/twp/pkg/mod/github.com/alecthomas/assert/v2@v2.4.1/assert.go:286 +0x294
  github.com/alecthomas/assert/v2.Equal[go.shape.*uint8]()
      /Users/twp/pkg/mod/github.com/alecthomas/assert/v2@v2.4.1/assert.go:51 +0x2b4
  github.com/twpayne/go-geos/geometry_test.TestNewGeometry()
      /Users/twp/src/github.com/twpayne/go-geos/geometry/geometry_test.go:197 +0x1cc
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x40

Previous write at 0x00c000138120 by goroutine 6:
  ??()
      -:0 +0x104cb6acc
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x14
  github.com/twpayne/go-geos.(*Geom).Destroy.func2()
      /Users/twp/src/github.com/twpayne/go-geos/geom.go:32 +0x34
  runtime.deferreturn()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/panic.go:477 +0x34
  github.com/twpayne/go-geos.(*Geom).finalize()
      /Users/twp/src/github.com/twpayne/go-geos/geom.go:339 +0x9c

Goroutine 106 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1648 +0x5e8
  testing.runTests.func1()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:2054 +0x80
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1595 +0x1b0
  testing.runTests()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:2052 +0x6e4
  testing.(*M).Run()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/testing/testing.go:1925 +0x9ec
  main.main()
      _testmain.go:61 +0x294

Goroutine 6 (running) created at:
  runtime.createfing()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/mfinal.go:163 +0x7c
  os.NewFile()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/os/file_unix.go:119 +0x80
  os.init()
      /opt/homebrew/Cellar/go/1.21.5/libexec/src/os/file.go:65 +0x350
==================
--- FAIL: TestNewGeometry (0.00s)
    testing.go:1465: race detected during execution of test
FAIL
FAIL    github.com/twpayne/go-geos/geometry     0.516s
FAIL
```

It is unclear which package contains the race condition.

--- a/geom.go
+++ b/geom.go
@@ -134,10 +134,6 @@ func (g *Geom) NearestPoints(other *Geom) [][]float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	s := C.GEOSNearestPoints_r(g.context.handle, g.geom, other.geom)
 	if s == nil {
 		return nil
@@ -227,10 +223,6 @@ func (g *Geom) RelatePattern(other *Geom, pat string) bool {
 	defer C.free(unsafe.Pointer(patCStr))
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSRelatePattern_r(g.context.handle, g.geom, other.geom, patCStr) {
 	case 0:
 		return false

--- a/geometry/geometry_test.go
+++ b/geometry/geometry_test.go
@@ -194,7 +194,7 @@ func TestNewGeometry(t *testing.T) {
 
 	actual, err := geometry.NewGeometryFromGeoJSON([]byte(`{"type":"Point","coordinates":[1,2]}`))
 	assert.NoError(t, err)
-	assert.Equal(t, expected, actual)
+	assert.Equal(t, expected, actual, assert.Exclude[*geos.Context]())
 
 	wkb, err := hex.DecodeString("0101000000000000000000f03f0000000000000040")
 	assert.NoError(t, err)

--- a/geommethods.go
+++ b/geommethods.go
@@ -104,10 +104,6 @@ func (g *Geom) Contains(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSContains_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -139,10 +135,6 @@ func (g *Geom) CoveredBy(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSCoveredBy_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -158,10 +150,6 @@ func (g *Geom) Covers(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSCovers_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -177,10 +165,6 @@ func (g *Geom) Crosses(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSCrosses_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -204,10 +188,6 @@ func (g *Geom) Difference(other *Geom) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSDifference_r(g.context.handle, g.geom, other.geom), nil)
 }
 
@@ -216,10 +196,6 @@ func (g *Geom) DifferencePrec(other *Geom, gridSize float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSDifferencePrec_r(g.context.handle, g.geom, other.geom, C.double(gridSize)), nil)
 }
 
@@ -228,10 +204,6 @@ func (g *Geom) Disjoint(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSDisjoint_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -247,10 +219,6 @@ func (g *Geom) Distance(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var distance float64
 	if C.GEOSDistance_r(g.context.handle, g.geom, other.geom, (*C.double)(&distance)) == 0 {
 		panic(g.context.err)
@@ -263,10 +231,6 @@ func (g *Geom) DistanceIndexed(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var distanceIndexed float64
 	if C.GEOSDistanceIndexed_r(g.context.handle, g.geom, other.geom, (*C.double)(&distanceIndexed)) == 0 {
 		panic(g.context.err)
@@ -279,10 +243,6 @@ func (g *Geom) DistanceWithin(other *Geom, dist float64) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSDistanceWithin_r(g.context.handle, g.geom, other.geom, C.double(dist)) {
 	case 0:
 		return false
@@ -314,10 +274,6 @@ func (g *Geom) Equals(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSEquals_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -333,10 +289,6 @@ func (g *Geom) EqualsExact(other *Geom, tolerance float64) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSEqualsExact_r(g.context.handle, g.geom, other.geom, C.double(tolerance)) {
 	case 0:
 		return false
@@ -352,10 +304,6 @@ func (g *Geom) FrechetDistance(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var frechetDistance float64
 	if C.GEOSFrechetDistance_r(g.context.handle, g.geom, other.geom, (*C.double)(&frechetDistance)) == 0 {
 		panic(g.context.err)
@@ -368,10 +316,6 @@ func (g *Geom) FrechetDistanceDensify(other *Geom, densifyFrac float64) float64 
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var frechetDistanceDensify float64
 	if C.GEOSFrechetDistanceDensify_r(g.context.handle, g.geom, other.geom, C.double(densifyFrac), (*C.double)(&frechetDistanceDensify)) == 0 {
 		panic(g.context.err)
@@ -399,10 +343,6 @@ func (g *Geom) HausdorffDistance(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var hausdorffDistance float64
 	if C.GEOSHausdorffDistance_r(g.context.handle, g.geom, other.geom, (*C.double)(&hausdorffDistance)) == 0 {
 		panic(g.context.err)
@@ -415,10 +355,6 @@ func (g *Geom) HausdorffDistanceDensify(other *Geom, densifyFrac float64) float6
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	var hausdorffDistanceDensify float64
 	if C.GEOSHausdorffDistanceDensify_r(g.context.handle, g.geom, other.geom, C.double(densifyFrac), (*C.double)(&hausdorffDistanceDensify)) == 0 {
 		panic(g.context.err)
@@ -447,10 +383,6 @@ func (g *Geom) Intersection(other *Geom) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSIntersection_r(g.context.handle, g.geom, other.geom), nil)
 }
 
@@ -459,10 +391,6 @@ func (g *Geom) IntersectionPrec(other *Geom, gridSize float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSIntersectionPrec_r(g.context.handle, g.geom, other.geom, C.double(gridSize)), nil)
 }
 
@@ -471,10 +399,6 @@ func (g *Geom) Intersects(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSIntersects_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -565,10 +489,6 @@ func (g *Geom) LargestEmptyCircle(other *Geom, tolerance float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSLargestEmptyCircle_r(g.context.handle, g.geom, other.geom, C.double(tolerance)), nil)
 }
 
@@ -665,10 +585,6 @@ func (g *Geom) Overlaps(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSOverlaps_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -692,10 +608,6 @@ func (g *Geom) Project(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return float64(C.GEOSProject_r(g.context.handle, g.geom, other.geom))
 }
 
@@ -704,10 +616,6 @@ func (g *Geom) ProjectNormalized(other *Geom) float64 {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return float64(C.GEOSProjectNormalized_r(g.context.handle, g.geom, other.geom))
 }
 
@@ -716,10 +624,6 @@ func (g *Geom) Relate(other *Geom) string {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	relateCStr := C.GEOSRelate_r(g.context.handle, g.geom, other.geom)
 	defer C.GEOSFree_r(g.context.handle, unsafe.Pointer(relateCStr))
 	return C.GoString(relateCStr)
@@ -730,10 +634,6 @@ func (g *Geom) RelateBoundaryNodeRule(other *Geom, bnr RelateBoundaryNodeRule) s
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	relateBoundaryNodeRuleCStr := C.GEOSRelateBoundaryNodeRule_r(g.context.handle, g.geom, other.geom, C.int(bnr))
 	defer C.GEOSFree_r(g.context.handle, unsafe.Pointer(relateBoundaryNodeRuleCStr))
 	return C.GoString(relateBoundaryNodeRuleCStr)
@@ -760,10 +660,6 @@ func (g *Geom) SharedPaths(other *Geom) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSSharedPaths_r(g.context.handle, g.geom, other.geom), nil)
 }
 
@@ -780,10 +676,6 @@ func (g *Geom) Snap(other *Geom, tolerance float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSSnap_r(g.context.handle, g.geom, other.geom, C.double(tolerance)), nil)
 }
 
@@ -800,10 +692,6 @@ func (g *Geom) SymDifference(other *Geom) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSSymDifference_r(g.context.handle, g.geom, other.geom), nil)
 }
 
@@ -812,10 +700,6 @@ func (g *Geom) SymDifferencePrec(other *Geom, gridSize float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSSymDifferencePrec_r(g.context.handle, g.geom, other.geom, C.double(gridSize)), nil)
 }
 
@@ -832,10 +716,6 @@ func (g *Geom) Touches(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSTouches_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false
@@ -867,10 +747,6 @@ func (g *Geom) Union(other *Geom) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSUnion_r(g.context.handle, g.geom, other.geom), nil)
 }
 
@@ -879,10 +755,6 @@ func (g *Geom) UnionPrec(other *Geom, gridSize float64) *Geom {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.GEOSUnionPrec_r(g.context.handle, g.geom, other.geom, C.double(gridSize)), nil)
 }
 
@@ -891,10 +763,6 @@ func (g *Geom) Within(other *Geom) bool {
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.GEOSWithin_r(g.context.handle, g.geom, other.geom) {
 	case 0:
 		return false

--- a/geommethods.go.tmpl
+++ b/geommethods.go.tmpl
@@ -40,10 +40,6 @@ func (g *Geom) {{ .name }}(other *Geom{{ range .extraArgs }}, {{ .name }} {{ .ty
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	return g.context.newGeom(C.{{ $geosFunction }}(g.context.handle, g.geom, other.geom{{ range .extraArgs }}, {{ .type | cType }}({{ .name }}){{ end }}), nil)
 }
 
@@ -71,10 +67,6 @@ func (g *Geom) {{ .name }}(other *Geom{{ range .extraArgs }}, {{ .name }} {{ .ty
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	switch C.{{ $geosFunction }}(g.context.handle, g.geom, other.geom{{ range .extraArgs }}, {{ .type | cType }}({{ .name }}){{ end }}) {
 	case 0:
 		return false
@@ -108,10 +100,6 @@ func (g *Geom) {{ .name }}(other *Geom{{ range .extraArgs }}, {{ .name }} {{ .ty
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
     {{- if .valueReturned }}
 	return float64(C.{{ $geosFunction }}(g.context.handle, g.geom, other.geom{{ range .extraArgs }}, {{ .type | cType }}({{ .name }}){{ end }}))
     {{- else }}
@@ -131,10 +119,6 @@ func (g *Geom) {{ .name }}(other *Geom{{ range .extraArgs }}, {{ .name }} {{ .ty
 	g.mustNotBeDestroyed()
 	g.context.Lock()
 	defer g.context.Unlock()
-	if other.context != g.context {
-		other.context.Lock()
-		defer other.context.Unlock()
-	}
 	{{ $varName }}CStr := C.{{ $geosFunction }}(g.context.handle, g.geom, other.geom{{ range .extraArgs }}, {{ .type | cType }}({{ .name }}){{ end }})
 	defer C.GEOSFree_r(g.context.handle, unsafe.Pointer({{ $varName }}CStr))
 	return C.GoString({{ $varName }}CStr)

